src/main/kotlin/org/rust/ide/typing/RustRawLiteralHashesBalancer.kt

```diff
@@ -0,0 +1,89 @@
+import com.intellij.psi.PsiFile
+import org.rust.lang.core.psi.RustTokenElementTypes.RAW_LITERALS
+import org.rust.lang.core.psi.impl.RustFile
+import org.rust.lang.core.psi.impl.RustRawStringLiteralImpl
+
+class RustRawLiteralHashesInserter : TypedHandlerDelegate() {
+    override fun beforeCharTyped(c: Char, project: Project, editor: Editor, file: PsiFile, fileType: FileType): Result {
+        if (file !is RustFile) return Result.CONTINUE
+
+        // We are only interested in handling typing '#'
+        if (c != '#') return Result.CONTINUE
+
+        val caretOffset = editor.caretModel.offset
+        if (!isValidOffset(caretOffset, editor.document.charsSequence)) return Result.CONTINUE
+
+        // Get token type current cursor is, we are using caretOffset - 1 in order to properly
+        // handle this case: r#""#<caret>, cases r<caret>#""# and <caret>r#""# will still work.
+        val highlighter = (editor as EditorEx).highlighter
+        val iterator = highlighter.createIterator(caretOffset - 1)
+        val (openHashes, closeHashes) = getHashesOffsets(iterator) ?: return Result.CONTINUE
+
+        // Now detect on which side of the literal we are, and insert hash on the other one
+        when (caretOffset) {
+            in openHashes -> editor.document.insertString(closeHashes.endOffset - 1, "#")
+            in closeHashes -> editor.document.insertString(openHashes.startOffset, "#")
+        }
+
+        return Result.CONTINUE
+    }
+}
+
+class RustRawLiteralHashesDeleter : RustEnableableBackspaceHandlerDelegate() {
+    private var offsets: Pair<TextRange, TextRange>? = null
+
+    override fun deleting(c: Char, file: PsiFile, editor: Editor): Boolean {
+        val caretOffset = editor.caretModel.offset
+        if (!isValidOffset(caretOffset, editor.document.charsSequence)) return false
+
+        val highlighter = (editor as EditorEx).highlighter
+        val iterator = highlighter.createIterator(caretOffset - 1)
+
+        // [getHashesOffsets] is O(n) (n = literal length), so do not evaluate it when it's not needed.
+        if (c != '#' || iterator.tokenType !in RAW_LITERALS) return false
+
+        // We have to compute offsets here, because we still have our '#' in document.
+        offsets = getHashesOffsets(iterator)
+        return offsets != null
+    }
+
+    override fun deleted(c: Char, file: PsiFile, editor: Editor): Boolean {
+        // We want caret offset before deletion!
+        val caretOffset = editor.caretModel.offset + 1
+        val (openHashes, closeHashes) = checkNotNull(offsets)
+
+        // Now detect on which side of the literal we are, and remove hash on the other one.
+        // Remember that offsets apply to literal before deletion!
+        when (caretOffset) {
+            in openHashes -> {
+                // -1 because left-closed ranges
+                // -1 because open hash was deleted some range is shifted do right by 1
+                // TODO: Why one more -1?
```

Comment:
Looks like hash ranges are one greater than they should be. for

```
fn main() {
    r###"aaa"###
}
```

I got

```
openHashes = (17,21)
closeHashes = (25,29)
```

That is, the ranges are of length 4, while looks like thy should be of the length 3. Is this a bug? If it is, then perhaps a test is missing?